File: ruby/linked_list/206_reverse_linked_list.rb
# Definition for singly-linked list.
class ListNode
  attr_accessor :val, :next

  def initialize(val = 0, nxt = nil)
    @val = val
    @next = nxt
  end

  def to_a
    result = []
    current = self
    until current.nil?
      result.append(current.val)
      current = current.next
    end
    result
  end
end

# @param {ListNode} head
# @return {ListNode}
def reverse_list(head)
  nil if head.nil?

  current = head
  prev = nil
  until current.nil?
    nxt = current.next
    current.next = prev

    prev = current
    current = nxt
  end

  return prev
end

def populate_list(values = [])
  return nil if values.empty?

  head = ListNode.new(values[0])
  current = head

  values[1..].each do |value|
    nxt = ListNode.new(value)
    current.next = nxt

    current = nxt
  end

  head
end

def test
  #Test cases

  # result: [5, 4, 3, 2, 1]
  puts reverse_list(populate_list([1,2,3,4,5])).to_a

  # result: []
  puts reverse_list(populate_list([])).to_a

  # result: [2, 1]
  puts reverse_list(populate_list([1, 2])).to_a
end

test

File: go/linked_list/206_reverse_linked_list.go
/**
 * Definition for singly-linked list.
 */
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	if head == nil {
		return head
	}

	var prev *ListNode
	current := head
	for current != nil {
		next := current.Next
		current.Next = prev

		prev = current
		current = next
	}

	return prev
}

func populateList(values []int) *ListNode {
	if len(values) == 0 {
		return nil
	}

	head := ListNode{values[0], nil}
	current := &head

	for _, value := range values[1:len(values)] {
		next := ListNode{value, nil}
		current.Next = &next
		current = current.Next
	}

	return &head
}

func printList(head *ListNode) {
	current := head
	for current != nil {
		fmt.Println(current.Val)
		current = current.Next
	}
}

func main() {
	// Test cases

	// result: [4, 3, 2, 1]
	printList(reverseList(populateList([]int{1, 2, 3, 4, 5})))

	// result: nil
	printList(reverseList(populateList([]int{})))

	// result: [2, 1]
	printList(reverseList(populateList([]int{1, 2})))
}
